src/USER-PHONON/dynamical_matrix.cpp

```diff
@@ -0,0 +1,533 @@
+    }
+}
+
+/* ---------------------------------------------------------------------- */
+
+void DynamicalMatrix::convert_units(const char *style)
+{
+    // physical constants from:
+    // http://physics.nist.gov/cuu/Constants/Table/allascii.txt
+    // using thermochemical calorie = 4.184 J
+
+    if (strcmp(style,"lj") == 0) {
+        error->all(FLERR,"Conversion Not Set");
+        //conversion = 1; // lj -> 10 J/mol
+
+    } else if (strcmp(style,"real") == 0) {
+        conv_energy = 418.4; // kcal/mol -> 10 J/mol
+        conv_mass = 1; // g/mol -> g/mol
+        conv_distance = 1; // angstrom -> angstrom
+
+    } else if (strcmp(style,"metal") == 0) {
+        conv_energy = 9648.5; // eV -> 10 J/mol
+        conv_mass = 1; // g/mol -> g/mol
+        conv_distance = 1; // angstrom -> angstrom
+
+    } else if (strcmp(style,"si") == 0) {
+        if (comm->me) error->warning(FLERR,"Conversion Warning: Multiplication by Large Float");
+        conv_energy = 6.022E22; // J -> 10 J/mol
+        conv_mass = 6.022E26; // kg -> g/mol
+        conv_distance = 1E-10; // meter -> angstrom
+
+    } else if (strcmp(style,"cgs") == 0) {
+        if (comm->me) error->warning(FLERR,"Conversion Warning: Multiplication by Large Float");
+        conv_energy = 6.022E12; // Erg -> 10 J/mol
+        conv_mass = 6.022E23; // g -> g/mol
+        conv_distance = 1E-7; // centimeter -> angstrom
+
+    } else if (strcmp(style,"electron") == 0) {
+        conv_energy = 262550; // Hartree -> 10 J/mol
+        conv_mass = 1; // amu -> g/mol
+        conv_distance = 0.529177249; // bohr -> angstrom
+
+    } else if (strcmp(style,"micro") == 0) {
+        if (comm->me) error->warning(FLERR,"Conversion Warning: Untested Conversion");
+        conv_energy = 6.022E10; // picogram-micrometer^2/microsecond^2 -> 10 J/mol
+        conv_mass = 6.022E11; // pg -> g/mol
+        conv_distance = 1E-4; // micrometer -> angstrom
+
+    } else if (strcmp(style,"nano") == 0) {
+        if (comm->me) error->warning(FLERR,"Conversion Warning: Untested Conversion");
+        conv_energy = 6.022E4; // attogram-nanometer^2/nanosecond^2 -> 10 J/mol
+        conv_mass = 6.022E5; // ag -> g/mol
+        conv_distance = 0.1; // angstrom -> angstrom
+
+    } else error->all(FLERR,"Units Type Conversion Not Found");
+
+}
+
+/* ---------------------------------------------------------------------- */
+
+void DynamicalMatrix::create_groupmap()
+{
+    //Create a group map which maps atom order onto group
+
+    int local_idx; // local index
+    int gid = 0; //group index
+    int nlocal = atom->nlocal;
+    int *mask = atom->mask;
+    bigint natoms = atom->natoms;
+    int *recv = new int[comm->nprocs];
+    int *displs = new int[comm->nprocs];
+    int *temp_groupmap = new int[natoms];
+
+    //find number of local atoms in the group (final_gid)
+    for (int i=1; i<=natoms; i++){
+        local_idx = atom->map(i);
+        if (mask[local_idx] & groupbit && local_idx < nlocal)
```

Comment:
This does not account for `local_idx < 0` for atoms that are not owned and will thus have illegal memory access. Try:
```
if ((local_idx >= 0) && (local_idx < nlocal) && mask[local_idx] & groupbit) 
```